openmetadata-service/src/main/java/org/openmetadata/service/security/auth/AzureAuthenticator.java

```diff
@@ -0,0 +1,457 @@
+      user.setEmail(email);
+      user.setName(name != null ? name : email);
+      user.setUpdatedAt(System.currentTimeMillis());
+      user.setUpdatedBy(email);
+      return userRepository.create(null, user);
+    } catch (Exception ex) {
+      throw new AuthenticationException("Failed to lookup user");
+    }
+  }
+
+  @Override
+  public void validateGroupAndRoleMapping(JWT jwt, String accessToken)
+      throws AuthenticationException {
+    try {
+      JWTClaimsSet claims;
+      if (!isSelfSignUpAvailable) {
+        return;
+      }
+
+      if (jwt instanceof SignedJWT signedJWT) {
+        claims = signedJWT.getJWTClaimsSet();
+      } else {
+        claims = jwt.getJWTClaimsSet();
+      }
+
+      List<String> azureGroups = claims.getStringListClaim("groups");
+      List<String> roles = claims.getStringListClaim("roles");
+      String userEmail = claims.getStringClaim("email");
+
+      if (userEmail == null) {
+        // Fallback to preferred_username if email claim is not present
+        userEmail = claims.getStringClaim("preferred_username");
+        if (userEmail == null) {
+          LOG.error("Neither email nor preferred_username claim found in token");
+          throw new AuthenticationException("Invalid token payload - no email identifier found");
+        }
+      }
+
+      String displayName = claims.getStringClaim("name");
+      if (displayName == null || displayName.trim().isEmpty()) {
+        displayName = userEmail;
+      }
+
+      User user = lookUserInProvider(userEmail, displayName);
+
+      // Handle Azure AD group mapping if enabled
+      if (isGroupMappingEnabled) {
+        LOG.debug("Processing group mapping for user {}", userEmail);
+        List<String> groupsToSync = (azureGroups != null) ? azureGroups : new ArrayList<>();
+        syncUserTeams(user, groupsToSync, accessToken);
+      }
+
+      // Handle Azure AD role mapping if enabled
+      if (isRoleMappingEnabled) {
+        LOG.debug("Processing role mapping for user {}", userEmail);
+        List<String> rolesToSync = (roles != null) ? roles : new ArrayList<>();
+        syncUserRoles(user, rolesToSync, accessToken);
+      }
+
+    } catch (AuthenticationException e) {
+      throw e;
+    } catch (Exception e) {
+      LOG.error("Failed to validate Azure JWT or sync user teams/roles", e);
+      throw new AuthenticationException("Azure JWT validation or team/role sync failed", e);
+    }
+  }
+
+  protected List<String> getAzureGroupHierarchyPath(String groupId, String accessToken)
+      throws IOException, InterruptedException {
+    List<String> hierarchy = new ArrayList<>();
+    String currentGroupId = groupId;
+
+    while (currentGroupId != null) {
+      // Get current group's info
+      HttpRequest groupRequest =
+          HttpRequest.newBuilder()
+              .uri(URI.create("https://graph.microsoft.com/v1.0/groups/" + currentGroupId))
+              .header("Authorization", "Bearer " + accessToken)
```

Comment:
static string with access token substitution is better than the contact like this. You can create final strings to substitute into it